java-checks/src/main/java/org/sonar/java/checks/ForLoopVariableTypeCheck.java

```diff
@@ -0,0 +1,130 @@
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    ForEachStatement actualStatement = (ForEachStatement) tree;
+    Type variableType = actualStatement.variable().type().symbolType();
+    Type collectionItemType = getCollectionItemType(actualStatement.expression());
+
+    if (collectionItemType != null && !isMostPreciseType(variableType, collectionItemType)) {
+      // Second pass: check if the variable is down-cast in the statement block
+      DownCastVisitor downCastVisitor = new DownCastVisitor(actualStatement, actualStatement.variable(), collectionItemType);
+      downCastVisitor.setContext(context);
+      downCastVisitor.scanTree(((ForEachStatement) tree).statement());
+    }
+  }
+
+  @CheckForNull
+  private static JavaType getCollectionItemType(ExpressionTree expression) {
+    if (((JavaType) expression.symbolType()).isParameterized()) {
+      ParametrizedTypeJavaType paramType = (ParametrizedTypeJavaType) expression.symbolType();
+      return paramType.substitution(paramType.typeParameters().get(0));
+    } else {
+      return null;
+    }
+  }
+
+  private static boolean isMostPreciseType(Type variableType, Type collectionItemType) {
+    if (collectionItemType instanceof WildCardType) {
+      return ((WildCardType) collectionItemType).isSubtypeOfBound((JavaType) variableType);
+    } else if (collectionItemType instanceof ParametrizedTypeJavaType) {
+      return ((ParametrizedTypeJavaType) collectionItemType).erasure().equals(variableType.erasure());
+    } else {
+      return variableType.equals(collectionItemType);
+    }
+  }
+
+  private class DownCastVisitor extends SubscriptionVisitor {
+
+    private final ForEachStatement forEachStatement;
+    private final VariableTree variable;
+    private final Type collectionItemType;
+    private boolean blockAlreadyFlagged;
+
+    private DownCastVisitor(ForEachStatement forEachStatement, VariableTree variable, Type collectionItemType) {
+      this.forEachStatement = forEachStatement;
+      this.variable = variable;
+      this.collectionItemType = collectionItemType;
+      this.blockAlreadyFlagged = false;
+    }
+
+    @Override
+    public List<Tree.Kind> nodesToVisit() {
+      return Collections.singletonList(Tree.Kind.TYPE_CAST);
+    }
+
+    @Override
+    public void visitNode(Tree tree) {
+      if (blockAlreadyFlagged) {
+        return;
+      }
+      ExpressionTree expression = ((TypeCastTree) tree).expression();
+      if (expression.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) expression).symbol().equals(variable.symbol())) {
+        ForLoopVariableTypeCheck.this.reportIssue(forEachStatement.variable().type(), String.format(PRIMARY_MESSAGE, variable.type().symbolType().name()),
+          getSecondaryLocations(), 0);
+        blockAlreadyFlagged = true;
+      }
+    }
+
+    @Override
+    public void scanTree(Tree tree) {
+      super.scanTree(tree);
+    }
```

Comment:
You could remove this override as it does not do anything

Reply:
It does something very subtle: it makes the method ``public`` instead of ``protected`` :)